d2/src/main/java/com/linkedin/d2/balancer/clients/RetryClient.java

```diff
@@ -254,16 +338,201 @@ public void onError(Throwable e)
 
     private boolean isRetryException(Throwable e)
     {
-      return ExceptionUtils.indexOfType(e, RetriableRequestException.class) != -1;
+      Throwable[] throwables = ExceptionUtils.getThrowables(e);
+
+      for (Throwable throwable: throwables)
+      {
+        if (throwable instanceof RetriableRequestException)
+        {
+          return !((RetriableRequestException) throwable).getDoNotRetryOverride();
+        }
+      }
+
+      return false;
     }
 
     /**
      * Retries a specific request.
      *
      * @param request Request to retry.
      * @param context Context of the retry request.
+     * @param numberOfRetryAttempts Number of retry attempts.
      * @return {@code true} if a request can be retried; {@code false} otherwise;
      */
-    public abstract boolean doRetryRequest(REQ request, RequestContext context);
+    public abstract boolean doRetryRequest(REQ request, RequestContext context, int numberOfRetryAttempts);
+  }
+
+  @ThreadSafe
+  private class ClientRetryTracker
+  {
+    private final int _aggregatedIntervalNum;
+    private final long _updateIntervalMs;
+    private final Clock _clock;
+    private final String _serviceName;
+
+    private final Object _counterLock = new Object();
+    private final Object _updateLock = new Object();
+
+    @GuardedBy("_updateLock")
+    private volatile long _lastRollOverTime;
+    private double _currentAggregatedRetryRatio;
+
+    @GuardedBy("_counterLock")
+    private final LinkedList<RetryCounter> _retryCounter;
+    private final RetryCounter _aggregatedRetryCounter;
+
+    private ClientRetryTracker(int aggregatedIntervalNum, long updateIntervalMs, Clock clock, String serviceName)
+    {
+      _aggregatedIntervalNum = aggregatedIntervalNum;
+      _updateIntervalMs = updateIntervalMs;
+      _clock = clock;
+      _serviceName = serviceName;
+
+      _lastRollOverTime = clock.currentTimeMillis();
+      _currentAggregatedRetryRatio = 0;
+
+      _aggregatedRetryCounter = new RetryCounter();
+      _retryCounter = new LinkedList<>();
+      _retryCounter.add(new RetryCounter());
+    }
+
+    public void add(boolean isRetry)
+    {
+      synchronized (_counterLock)
+      {
+        if (isRetry)
+        {
+          _retryCounter.getLast().addToRetryRequestCount(1);
+        }
+
+        _retryCounter.getLast().addToTotalRequestCount(1);
+      }
+      updateRetryDecision();
+    }
+
+    public void rollOverStats()
+    {
+      // rollover the current interval to the aggregated counter
+      synchronized (_counterLock)
+      {
+        RetryCounter intervalToAggregate = _retryCounter.getLast();
+        _aggregatedRetryCounter.addToTotalRequestCount(intervalToAggregate.getTotalRequestCount());
+        _aggregatedRetryCounter.addToRetryRequestCount(intervalToAggregate.getRetryRequestCount());
```

Comment:
Minor: if you always operate on total and retry count together, maybe you can merge these 2 lines into 2 method? Will leave the decision to you

Reply:
I feel like the current API of RetryCounter is clearer by separating out total request with retry request. Here I need to call the two methods separately https://github.com/linkedin/rest.li/pull/484/files#diff-fb1d5cedc2fd04937e39f9188139915da2a2ae4aa9c736de7503973cc85c4b6eR419